test/integration/autonomous_upstream.cc

```diff
@@ -0,0 +1,77 @@
+#include "test/integration/autonomous_upstream.h"
+
+namespace Envoy {
+
+namespace {
+
+void HeaderToInt(const char header_name[], int32_t& return_int, Http::TestHeaderMapImpl& headers) {
+  std::string header_value = headers.get_(header_name);
+  if (!header_value.empty()) {
+    return_int = atoi(header_value.c_str());
+  }
+}
+
+} // namespace
+
+const char AutonomousStream::RESPONSE_SIZE_BYTES[] = "response_size_bytes";
+const char AutonomousStream::EXPECT_REQUEST_SIZE_BYTES[] = "expect_request_size_bytes";
+const char AutonomousStream::RESET_AFTER_REQUEST[] = "reset_after_request";
+
+// For now, assert all streams which are started are completed.
+// Support for incomplete streams can be added when needed.
+AutonomousStream::~AutonomousStream() { RELEASE_ASSERT(complete()); }
+
+// By default, automatically send a response when the request is complete.
+void AutonomousStream::setEndStream(bool end_stream) {
+  FakeStream::setEndStream(end_stream);
+  if (end_stream) {
+    sendResponse();
+  }
+}
+
+// Check all the special headers and send a customized response based on them.
+void AutonomousStream::sendResponse() {
+  Http::TestHeaderMapImpl headers(*headers_);
+
+  int32_t request_body_length = -1;
+  HeaderToInt(EXPECT_REQUEST_SIZE_BYTES, request_body_length, headers);
```

Comment:
Yeah, that's what I mean by plumbing through - we'd have to add an API for each to AutonomousUpstream, AutonomousHttpConnection and AutonomousStream

The other nice benefit of doing this in headers is it ties the requests to responses cleanly.  Google style if we said "every 5th stream at the backend gets reset" it would be tricky associating the client streams to backend streams, since there's no consistent ordering guarantee.  Here because we do it in the stream headers, you can easily ensure the specific stream which requested reset will be reset.

Reply:
FWIW I like the header approach. I think it's really easy to reason about and easy to use.